allennlp/data/fields/array_field.py

```diff
@@ -30,23 +41,28 @@ def as_tensor(self,
                      for i in range(len(padding_lengths))]
 
         return_array = numpy.ones(max_shape, "float32") * self.padding_value
-
         # If the tensor has a different shape from the largest tensor, pad dimensions with zeros to
         # form the right shaped list of slices for insertion into the final tensor.
         slicing_shape = list(self.array.shape)
         if len(self.array.shape) < len(max_shape):
             slicing_shape = slicing_shape + [0 for _ in range(len(max_shape) - len(self.array.shape))]
         slices = [slice(0, x) for x in slicing_shape]
-        return_array[slices] = self.array
+        # torch seems to have trouble with this, so we cast to numpy
+        if self.array_type == "numpy":
+            return_array[slices] = self.array
+        else:
+            return_array[slices] = self.array.cpu().numpy()
         tensor = torch.from_numpy(return_array)
         return tensor if cuda_device == -1 else tensor.cuda(cuda_device)
```

Comment:
if I understand this correctly, in the tensor case, if your tensors are on the GPU (which is how I assume you'd want them in the GPU case), then `as_tensor` moves them to the CPU, convert them to numpy arrays, and then move them back to the GPU?!

Reply:
In the Tensor case, it’s actually more likely (at least for my application) that the tensors are on CPU —- there isn’t enough GPU memory to store them all there. This is still a fair criticism, though. The fancy indexing didn’t work out of the box in pytorch, so I’ll see if I can find a workaround.

Reply:
If they are going to be on the CPU anyway, we should just close these PRs and you should use an ArrayField - torch tensors will share memory with numpy arrays, so calls to and from numpy will be very fast:

```
(allennlp) markn@markn ~ $ python -mtimeit -s "import torch" "torch.from_numpy(torch.randn([20, 30, 1000]).numpy())"
100 loops, best of 3: 4.86 msec per loop
```
A full round trip for a pretty large tensor takes 5/1000ths of a second. 

I thought the reason for this PR was because you had some tensors transitively on the GPU, which you wanted to keep there. If not, perhaps underoptimise before you over optimise 😃

Reply:
You're right, that solution seems to be the simplest and just work. Thanks @DeNeutoy